### ISSUE TITLE
Use type_info in equality comparison of any_scheduler 

### ISSUE THREAD

**Issue by maikel** [NVIDIA/stdexec#797]
Currently, we compare pointers to vtable objects in static storage duration to determine if the underlying scheduler types are equal. I think, that is not correct.

Instead we should use something like type_info.

**Comment by villevoutilainen** [comment]
Why? The technique of comparing pointers works in e.g. the implementation of std::any,  which compares the address of the stored accessor function to the one that would be used to convert the any to Foo, so if the accessor is not the accessor of Foo, the conversion will fail. That works fine. Why do you think we need type_info here?

**Comment by jrhemstad** [comment]
> Why? The technique of comparing pointers works in e.g. the implementation of std::any, which compares the address of the stored accessor function to the one that would be used to convert the any to Foo, so if the accessor is not the accessor of Foo, the conversion will fail. That works fine. Why do you think we need type_info here?

fwiw, this approach doesn't work on Windows because each DLL will have it's own copy of the vtable and the pointers will be different. Granted, stdexec doesn't care about Windows right now, but one day it probably will.

**Comment by jrhemstad** [comment]
> ..but these are not actual vtables.

Indeed, but the static storage duration objects will still be unique in each DLL. In other words, singletons aren't shared across DLL's.

**Comment by maikel** [comment]
> > ..but these are not actual vtables.
> 
> Indeed, but the static storage duration objects will still be unique in each DLL. In other words, singletons aren't shared across DLL's.

This. I just wanted to keep a note about a possible deficiency. My personal motivation and use case for senders includes its usage in interfaces across shared libraries. That's why I happily implemented the type erasure in the first place.

**Comment by villevoutilainen** [comment]
> > ..but these are not actual vtables.
> 
> Indeed, but the static storage duration objects will still be unique in each DLL. In other words, singletons aren't shared across DLL's.

Even if they're inline?

**Comment by ericniebler** [comment]
> Even if they're inline?

I suppose that depends on whether the dynamic linker knows anything about inline variables. I'm guessing it doesn't, but I would need to test to be certain.
